### ISSUE TITLE
NeoVim fork?

### ISSUE THREAD

**Issue by herrbischoff** [macvim-dev/macvim#47]
Would you consider creating a fork of MacVim based on the NeoVim project instead of Vim? I believe that this is the future of Vim development and would love to see a more performant, more modern implementation of Vim come to life MacVim-style.

**Comment by eaoliver** [comment]
Just out of curiosity, why is it the future of Vim development?

**Comment by ChrisHonniball** [comment]
> Is it ready to download and run now with all the features?
> No. Although some features are a work in progress, Neovim isn't at a stable point. Using Neovim should be done with caution as things may change.
> 
> When will a beta be available?
> When it's ready.

Until these items change I wouldn't do change MacVim to this. I would however keep up with vim updates better as in #46.

**Comment by herrbischoff** [comment]
@eaoliver: To not repeat what others have said already (and better than I could), here are a couple of links to better illustrate my opinion. Emphasis is on *opinion*, as obviously there is never an absolute wrong or right. Ideally, I'd just like to kick off a discussion about it.

* <http://geoff.greer.fm/2015/01/15/why-neovim-is-better-than-vim/>
* <https://github.com/neovim/neovim/wiki/Introduction>
* <http://vimcasts.org/blog/2014/03/support-neovim/>

**Comment by cserb** [comment]
@herrbischoff I just found this https://github.com/rogual/neovim-dot-app

**Comment by herrbischoff** [comment]
@cserb Thanks, but the performance of that GUI is unfortunately abysmal.

**Comment by RichardDooling** [comment]
You can supposedly get "true colors" with the testing builds of iTerm using NeoVim but to me the colors still look weird. If I knew more I'd probably know why. Also scrolling always seems jerky to me running Vim in terminal or experimenting with NeoVim in terminal. So for now I like MacVim. I haven't messed with tmux, which looks like another fascinating distraction.

**Comment by justinmk** [comment]
It doesn't really make sense to fork MacVim, I don't think, unless there's a significant amount of code that would be worth re-using. If https://github.com/rogual/neovim-dot-app is slow, it's because it's a very new and probably not at all optimized. Lots of low-hanging fruit, and forking MacVim would have to deal with those same issues anyways.

**Comment by Mudox** [comment]
Neovim's idea for its UI clients -- [Neovim Smart UI Protocol](http://tarruda.github.io/articles/neovim-smart-ui-protocol/)
I think it's better to wait until this low level architecture is build up.

**Comment by splhack** [comment]
https://github.com/rogual/neovim-dot-app

**Comment by johnnyutahh** [comment]
Any recent thoughts on this topic (porting MacVim to a Neovim "core")?

**Comment by ychin** [comment]
This probably will not happen in foreseeable future. It's a pretty significant amount of work to port MacVim to Neovim, in particular because there are a lot of modifications to Vim source code to allow for MacVim to happen. Porting to Neovim means redoing all those work *and* maintaining a cross-compatible code base. While Neovim is more designed for plugins and therefore should require less intrusive code forking, it still means we have to write another layer to talk to it, not to mention certain MacVim features likely still require code modifications on Neovim.

Long story short, unless I see a compelling reason to, or someone works on it with a solid architecture to maintain compatibility with existing MacVim implementation, it's very unlikely to happen.

**Comment by ychin** [comment]
So I'm in the process of revamping our IPC (Inter-process Communication) structures (MacVim is using an old deprecated tech called Distributed Objects), and been investigating options to move away from it (#1157). It's very likely that I will end up using Apple's XPC (MacVim's goal has always been to integrate as much as possible using Apple built-in APIs rather than generic third-party dependencies), but it does raise the prospect of refactoring the IPC to be less interdependent, just as a general good software engineering practice. The current IPC system was put together in kind of an ad-hoc fashion over the years and deserve some love to make it more maintainable.

Because of that, I'm revisiting potentially supporting Neovim. Given that Neovim already natively supports a multi-process embedding model, the work of supporting Neovim will be significantly less than the initial work on MacVim (which involved writing its own inter-process communication pipelines). We could treat the msgpack UI as another pipe with similar APIs as the one MacVim has for talking to Vim (they are similar because, well, Neovim evolved from Vim). I also think there are enough sharable features (I mean, both are basically Vim-like so to speak with similar requirements) that it makes some sense to share code.

---

Now, just a general question for people on this thread though: Is such a project/fork/work still appreciated or wanted? There are now lots of Neovim GUIs, including Mac-specific ones like VimR, or cross-platform ones like Neovim-Qt or Neovide. I just want to gauge general interests before I start spending too much time on it, as even though I said "less work", it is very very far from "no work".

The other thing is: MacVim relies extensively on upstream modification to achieve certain features, e.g. mapping `<SwipeLeft>` on trackpad to a command in Vim. Certain MacVim features like scroll bars also come for free from Vim and in Neovim it's more a "fend for yourself" model where the UI has to implement the logic themselves. So even if I get something working, for a while you won't get some of the features from MacVim today and there won't be feature parity. I may also need to fork Neovim for some of the MacVim work so I'm not sure if it will be able to just plug in to a generic Neovim instance.

I can at least guarantee that the app will be signed and notarized (which a lot of Neovim GUIs aren't) and works like a native Mac app :)

---

Note that this is just a thought for now and gauging interests, and not a promise to work on it. My primary responsibility is MacVim so that would be what takes most of my time and it wouldn't be fair to existing users if I drop off just to add a feature that they are not interested in. This is mostly an intellectual curiosity for me right now. For now I will probably just try to structure MacVim in a way that makes less assumptions of the backend (which is a good idea in general), just to leave the door open for later.

There are also enough differences between Vim and Neovim (in feature set and internal APIs) that will make this not as straightforward as I claimed above. It will kind of double my work in staying up-to-date with upstream changes (I currently read every Vim commit and having to do that with Neovim would be annoying lol). Maintainence will also be a bit painful with maintaining essentially two versions of the same app with two separate upstream to track.

Anyway comments are welcome. Some spices are good but I prefer no flames please. I will likely keep using Vim in the foreseeable future myself but I know Neovim is a popular alternative for some and I'm not philosophically opposed to a Neovim port (if it involved zero work).

**Comment by ychin** [comment]
> **I value MacVim tremendously; where can I contribute?** I'm a massive fan of MacVim. Thank you thank you @ychin for making and supporting it over the years, decades. It enabled me to avoid adopting/adapting any VSCode-like IDE as a primary/necessary swdev tool, which would have been painful. (An aside, to this end: is https://opencollective.com/macvim/contribute the best place to monetarily contribute? I'm long overdue to do so.)

Thanks! I do have to point out that I have been maintainer since 2018. Björn Winckler was the original author of MacVim and there were a couple maintainers in between us. Both https://github.com/sponsors/macvim-dev and  https://opencollective.com/macvim will work for me, so pick whatever works.

> * @ychin's [motivation for Neovim integration](https://github.com/macvim-dev/macvim/issues/47#issuecomment-1483950233) seems to be: it's the easiest way to introduce the next design/features (in this case: for IPC, if I'm reading from @ychin's above post correctly?)
> 
> * any additional motivations?

Well, yes, the biggest difference between how MacVim talks to Vim and a normal Neovim GUI talks to nvim is the IPC mechanism. Now, I don't really want to use msgpack (what Neovim uses) for how MacVim talks to Vim, and will likely use Apple's XPC, but it's just a good time to revisit this topic as I could design the system to be a little more IPC-system agnostic (note that XPC and a stream-based msgpack communication system are still *quite* different as underlying technology). I have also taken a look at different Neovim GUIs and the API before (for references) and it seemed doable in the long run.

Otherwise it's just that Neovim is a popular alternative to Vim and I'm not opposed to bringing a macOS-native GUI to it. I have used a few Neovim GUIs before and I think very few of them are really developed natively on a macOS and don't really pass the first bars of what I consider to be macOS-native in terms of feel and integration. Otherwise it's mostly an intellectual curiosity and therefore I want to justify with a need first.

> can you (@ychin) outline or highlight, to some degree, what might be lost in the "lack of feature parity" for initial MacVim-to-Neovim port/fork/whatever?

It's hard to give a full list. Some MacVim-specific features include: trackpad gesture mapping, scroll bars, autocmd for dark mode. Vim-specific features: `:gui` / `:hardcopy` command. Some builtin MacVim commands like autocmd for dark modes. A lot of these can be implemented via other ways but will take some work. Basically, assume that in a lot of ways it won't be better than a smart terminal in the beginning just to get things started.

There are things that I don't know until I investigate.

> can someone describe the high-level differences between Vim and Neovim?

This is a long topic and I recommend you just Google it. I'm saying this because if you don't know the difference then maybe it doesn't really matter to you :). Also, this topic has been debated ad nauseam in a lot of venues already. One key aspect is that Neovim is designed to be embedded in a UI rather than comes with it own GUI implementation. This both makes it easy to embed, and make it hard to get certain features "out of the box".

> can someone outline/highlight how MacVim is currently designed? eg: Vim "core" with a macOS-only GUI?

MacVim runs a forked version of Vim with custom modification (to support a macOS GUI and unique features like trackpad gestures). The actual app has a single top-level MacVim app process that hosts and talk to the Vim processes inside (each macOS window is a separate Vim process). This is similar to how a lot of Neovim GUIs work, except that most of them run unmodified Neovim instances and just launch it using `--headless` or `--embed` flags.

> will a MacVim-with-Neovim fork (?) require supporting 2 "vim cores" (Vim and Neovim)... or would the plan be to stop Vim one day and start Neovim the next?

No, MacVim will be *Vim*-first, and Neovim-second (if Neovim support is added) in any foreseeable future that I can see, unless the user count for MacVim has dropped to such a small amount I can count with one hand. The way I can see it happens is if Neovim support is added, and user count seems promising, then there could be more work done to make it better, re-implement certain missing features, and potentially add unique Neovim features (that take advantage of externalized UI), and potentially loop in more developers (which I don't think would be hard). If I somehow decide to switch to maintain a Neovim fork full time I will find a maintainer replacement for MacVim first.

**Comment by johnnyutahh** [comment]
> After doing some minimal research, it's likely I will be using, possibly even switching to, Neovim in the not-distant future. 

I'll also be looking into https://helix-editor.com (which is probably no surprise to some):

1. Excerpt from Helix.com: "If Neovim is the modern Vim, then Helix is post-modern."
2. [Helix FAQ: Is a Vi/Vim keymap planned?](https://github.com/helix-editor/helix/wiki/FAQ#is-a-vivim-keymap-planned) 

Helix seems like a signifcant departure from Vim/Neovim. I suspect it's "unthinkable" to consider a MacVim-Helix integration... but I'm massively and ignorantly guessing. I suspect many people here have far more understanding of this than I currently do.

**Comment by ychin** [comment]
> Helix seems like a signifcant departure from Vim/Neovim. I suspect it's "unthinkable" to consider a MacVim-Helix integration... but I'm massively and ignorantly guessing. I suspect many people here have far more understanding of this than I currently do.

No, even porting MacVim to Neovim is going to be a lot of work and something that I can't promise, so MacHelix is probably not something I have in mind. Someone can of course fork the code if they so wish, but I think they should just take the code as reference as write a new project from scratch (MacVim is quite old with a lot of curious hacks just to get it working in the early days and not the best code base to start off with for a new project).

**Comment by pusewicz** [comment]
Oh man, I was about to create an issue about this. It would be _fantastic_ if MacVim supported NeoVim. Lack of a polished, Mac-like UI is the only thing holding me back from the switch. VimR, while a great attempt, simply does not cut it.

**Comment by herrbischoff** [comment]
Interesting to see activity regarding Neovim after 8 years. I'm still all for it.

After adjusting to the Lua config environment, I absolutely don't want to go back to VimL. Treesitter support is another big one. Plus, sane defaults for a new century, as opposed to being anchored in 1989. Both approaches have merit of course. Personally, I don't need Amiga and AIX support any more.

I recently took stock of the state of Neovim GUIs (blog post in progress) and the result is sobering. The only one worth considering in my opinion is Neovide, a Rust implementation with minimal cruft. It has repeated major issues starting up at all though. A MacVim-quality GUI is still sorely needed, whichever form it may take in the end.

**Comment by 9mm** [comment]
> Thank you @ychin for [your thorough reply](https://github.com/macvim-dev/macvim/issues/47#issuecomment-1485947285), it's much appreciated.
> 
> After doing some minimal research, it's likely I will be using, possibly even switching to, Neovim in the not-distant future. Features like Neovim's LSP (eg: https://rsdlt.github.io/posts/rust-nvim-ide-guide-walkthrough-development-debug/) offer tremendous value. When I do, I'll be (re)testing the Neovim macOS GUIs, and I'll update here how much said GUIs are lacking (from my initial experience, at least) compared to MacVim.
> 
> An aside, and a point of note in my research: [Neovim is "the biggest love" for SO's 2022 IDE survey](https://survey.stackoverflow.co/2022/#section-most-loved-dreaded-and-wanted-integrated-development-environment). These SO developer surveys have served me well in the past, and I'm further motivated because of this result to explore Neovim.

I can answer you here... I've been using Macvim for 10+ years now (or around there), and it's truly 10/10.

I want to switch to Neovim _so badly_ but the GUI's are absolutely abysmal. IMO, the entire point of a GUI is to gain extra native support of the OS... after all, that's the entire point. Otherwise I would just use terminal neovim in iTerm.

I've tried all of them, VimR, Neovide, etc. They simply just don't properly do it the way Macvim did. For instance, command+Q doesnt even work in Neovide, it just exits the buffer without even asking to save. They have all these hacky workarounds with separate apps to try to get command keys.

I tried VimR a year ago, and I got kind-of sort-of close, but I guess that has a built in version of neovim thats always behind, and even though i got it close, it still didn't do many things properly like MacVim did.

I realize it's a lot of work though, and probably never would happen. It is depressing because I have to relearn a bunch of muscle memory now trying to switch to a less ideal setup

**Comment by swburk** [comment]
I switched full-time to Neovim about a year ago, enticed by built-in LSP support. I mostly use Neovim in the terminal because none of the existing GUIs are close to MacVim's level of simplicity and polish. I still use MacVim and maintain my Vim config strictly because it's often convenient to have my editor as a separate app I can quit, hide, and CMD + Tab to rather than getting lost in my sea of other terminal windows.

I don't think I have much to add to this conversation technically, but I would be ELATED if MacVim added support for Neovim. I simply wanted to post to cast my vote in favor of this.
